### PR TITLE
Add cacheable_unary interop test

### PR DESCRIPTION
**Pull request by ericgribkoff**
This PR has two commits:

1. Remove the `grpc-payload-bin=` prefix from outgoing GET requests. This brings the implementation inline with C++ and our test server behind GFE.
2. Add the cacheable_unary interop test, which verifies that GET requests sent to a server behind a caching proxy are correctly sent and responses cached.

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java [hunk 1]
```diff
@@ -128,8 +128,8 @@ public void writeHeaders(Metadata headers, byte[] requestPayload) {
       AsciiString httpMethod;
       if (get) {
         // Forge the query string
-        defaultPath = new AsciiString(defaultPath + "?" + GRPC_PAYLOAD_BIN_KEY + "="
-            + BaseEncoding.base64().encode(requestPayload));
+        defaultPath =
+            new AsciiString(defaultPath + "?" + BaseEncoding.base64().encode(requestPayload));
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Add a TODO for adding back the key.

File: interop-testing/src/test/java/io/grpc/testing/integration/TestCasesTest.java [hunk 2]
```diff
@@ -54,11 +54,27 @@ public void unknownStringThrowsException() {
...
+      "client_streaming",
+      "server_streaming",
+      "ping_pong",
+      "empty_stream",
+      "compute_engine_creds",
+      "service_account_creds",
+      "jwt_token_creds",
+      "oauth2_auth_token",
+      "per_rpc_creds",
+      "cacheable_unary",
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/TestCasesTest.java]: This is not in the right order. It appears it should be under empty_unary to match the interop spec. It needs to be fixed elsewhere as well.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 3]
```diff
@@ -1330,6 +1333,67 @@ public void perRpcCreds(String jsonKey, InputStream credentialsStream, String oa
     oauth2AuthToken(jsonKey, credentialsStream, oauthScope);
   }
 
+  /** Sends a cacheable unary rpc using GET. Requires that the server is behind a caching proxy. */
+  public void cacheableUnary() {
+    // Build method descriptor manually to set idempotent and safe to true.
+    io.grpc.MethodDescriptor<
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Why all this stuff? Why isn't CacheableUnaryCall in the proto? If it were, then I'd expect `TestServiceGrpc.METHOD_CACHEABLE_UNARY_CALL.toBuilder().setSafe(true).build()` here.

If we're going to have all this goo here, it seems at least we should remove all the unnecessary fully-qualified names (like io.grpc before MethodDescriptor).

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 4]
```diff
@@ -1330,6 +1333,67 @@ public void perRpcCreds(String jsonKey, InputStream credentialsStream, String oa
...
+                .build();
+    // Set fake user IP since some proxies (GFE) won't cache requests from localhost.
+    Metadata.Key<String> userIpKey = Metadata.Key.of("x-user-ip", Metadata.ASCII_STRING_MARSHALLER);
+    Metadata metadata = new Metadata();
+    metadata.put(userIpKey, "1.2.3.4");
+    Channel channelWithUserIpKey =
+        ClientInterceptors.intercept(channel, MetadataUtils.newAttachHeadersInterceptor(metadata));
+    SimpleRequest requests1And2 =
+        SimpleRequest.newBuilder()
+            .setResponseSize(1)
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Why is this requesting response details if it is just going to be thrown away? It doesn't say this should be sent in the interop spec.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 5]
```diff
@@ -1330,6 +1333,67 @@ public void perRpcCreds(String jsonKey, InputStream credentialsStream, String oa
...
+                    .setBody(ByteString.copyFromUtf8(String.valueOf(System.nanoTime()))))
+            .build();
+
+    SimpleResponse response1 =
+        ClientCalls.blockingUnaryCall(
+            channelWithUserIpKey, safeCacheableUnaryCallMethod, CallOptions.DEFAULT, requests1And2);
+    SimpleResponse response2 =
+        ClientCalls.blockingUnaryCall(
+            channelWithUserIpKey, safeCacheableUnaryCallMethod, CallOptions.DEFAULT, requests1And2);
+    SimpleResponse response3 =
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: This request is not in the interop test specification.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java (hunk 3 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: My mistake. I didn't know toBuilder() existed. Fixed.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java (hunk 4 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: The interop spec is going to be changed to allow validating that the server received the payload. Removed for now.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java (hunk 5 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: The interop test specification needs updating, which I was holding off on to do along with adding the verificiation of the payload as mentioned above. See https://github.com/grpc/grpc/blob/d93a8bead39ab112b808a248adb4576e4c9ce766/test/cpp/interop/interop_client.cc#L921 for the desired behavior.

File: interop-testing/src/test/java/io/grpc/testing/integration/TestCasesTest.java (hunk 2 above)
Comment (ericgribkoff) [inline interop-testing/src/test/java/io/grpc/testing/integration/TestCasesTest.java]: I didn't realize these were ordered according to the interop spec. Fixed.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 1 above)
Comment (ericgribkoff) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Done

File: testing-proto/src/main/proto/io/grpc/testing/integration/test.proto [hunk 6]
```diff
@@ -47,6 +47,11 @@ service TestService {
   // One request followed by one response.
   rpc UnaryCall(SimpleRequest) returns (SimpleResponse);
 
+  // One request followed by one response. Response has cache control
```
Comment (ejona86) [inline testing-proto/src/main/proto/io/grpc/testing/integration/test.proto]: Normally it is good to briefly sum up the important parts from the spec for the service implementation. Namely, that the request is ignored and the response is always different, except for caching based on the request.

Comment (ericgribkoff) [inline testing-proto/src/main/proto/io/grpc/testing/integration/test.proto]: This behavior is set to change soon, and the interop test protos have already diverged enough between here and grpc/grpc. Will update this comment once I add these files into grpc-proto and have a standardized location for the 
"ground truth".

**Pull request merged**